exonum-java-binding/core/rust/src/storage/proof_map_index.rs

```diff
@@ -12,49 +12,74 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use jni::{
-    objects::{JClass, JObject, JString},
-    sys::{jboolean, jbyteArray, jobject, jsize},
-    JNIEnv,
-};
-
-use std::{panic, ptr};
-
-use exonum::crypto::Hash;
 use exonum_merkledb::{
-    access::FromAccess,
+    access::{FromAccess, RawAccess},
     proof_map_index::{
-        MapProof, ProofMapIndexIter, ProofMapIndexKeys, ProofMapIndexValues, ProofPath,
-        PROOF_MAP_KEY_SIZE,
+        ProofMapIndexIter, ProofMapIndexKeys, ProofMapIndexValues, PROOF_MAP_KEY_SIZE,
     },
-    Fork, IndexAddress, ObjectHash, ProofMapIndex, Snapshot,
+    Fork, IndexAddress, ObjectHash, ProofMapIndex, RawProofMapIndex, Snapshot,
 };
-
+use exonum_proto::ProtobufConvert;
 use handle::{self, Handle};
+use jni::{
+    objects::{GlobalRef, JClass, JMethodID, JObject, JString},
+    sys::{jboolean, jbyteArray, jobject, jobjectArray, JNI_FALSE},
+    JNIEnv,
+};
+use protobuf::Message;
 use storage::{
-    db::{Value, View, ViewRef},
+    db::{Key, Value, View, ViewRef},
     PairIter,
 };
 use utils;
 use JniResult;
 
-type Key = [u8; PROOF_MAP_KEY_SIZE];
-type Index<T> = ProofMapIndex<T, Key, Value>;
+use std::{panic, ptr};
+
+type RawKey = [u8; 32];
+
+enum Index<T>
+where
+    T: RawAccess,
+{
+    Raw(RawProofMapIndex<T, RawKey, Value>),
+    Hashed(ProofMapIndex<T, Key, Value>),
+}
 
 const JAVA_ENTRY_FQN: &str = "com/exonum/binding/core/storage/indices/MapEntryInternal";
-const MAP_PROOF_ENTRY: &str = "com/exonum/binding/common/proofs/map/MapProofEntry";
-const MAP_ENTRY: &str = "com/exonum/binding/common/collect/MapEntry";
-const UNCHECKED_FLAT_MAP_PROOF: &str = "com/exonum/binding/common/proofs/map/UncheckedFlatMapProof";
-const UNCHECKED_FLAT_MAP_PROOF_SIG: &str =
-    "([Lcom/exonum/binding/common/proofs/map/MapProofEntry;[Lcom/exonum/binding/common/collect/MapEntry;[[B)Lcom/exonum/binding/common/proofs/map/UncheckedFlatMapProof;";
-const BYTE_ARRAY: &str = "[B";
 
 enum IndexType {
     SnapshotIndex(Index<&'static dyn Snapshot>),
     ForkIndex(Index<&'static Fork>),
 }
 
-type Iter<'a> = PairIter<ProofMapIndexIter<'a, Key, Value>>;
+enum Iter<'a> {
+    Raw(PairIter<ProofMapIndexIter<'a, RawKey, Value>>),
+    Hashed(PairIter<ProofMapIndexIter<'a, Key, Value>>),
+}
+
+enum KeysIter<'a> {
+    Raw(ProofMapIndexKeys<'a, RawKey>),
+    Hashed(ProofMapIndexKeys<'a, Key>),
+}
+
+// For easy conversion to RawKey.
+trait ToRawKey {
+    fn to_raw(&self) -> RawKey;
+}
+
+impl ToRawKey for Key {
```

Comment:
Really confusing implementation, because `Key == Vec<u8>` and we use it like
```rust
let key: Vec<u8> = env.convert_byte_array(key);
let raw_key = key.to_raw();
```

Reply:
Within a single method, Java passes us a byte array that can be one of two native types. We always have to convert it to the `Vec<u8>`, but in case of raw PMI we also `convert` it to its key format. This trait is just a substitution of explicit `convert_to_raw(key)`.

Reply:
Will implementing `From<Vec<u8>> for RawKey` work?

Reply:
`From` (1) consumes original value (probably, not a big deal) and (2) must never fail conversion (but OTOH we're sure conversion will never happen for other than 256 bit keys).